sphinx/environment/adapters/toctree.py

```diff
@@ -315,19 +316,23 @@ def _toctree_entry(
             # empty toc means: no titles will show up in the toctree
             logger.warning(__('toctree contains reference to document %r that '
                               "doesn't have a title: no link will be generated"),
-                           ref, location=toctreenode)
-    except KeyError:
-        # this is raised if the included file does not exist
+                           ref, location=toctreenode, type='toc', subtype='no_title')
+    except KeyError as err:
+        # It may happen only after `directives/other.py` parsing when
+        # the included document changed/disappeared. Thus, it is a not correctable warning.
         ref_path = env.doc2path(ref, False)
         if excluded(ref_path):
             message = __('toctree contains reference to excluded document %r')
+            subtype = 'excluded'
         elif not included(ref_path):
             message = __('toctree contains reference to non-included document %r')
+            subtype = 'not_included'
         else:
             message = __('toctree contains reference to nonexisting document %r')
+            subtype = 'not_readable'
 
-        logger.warning(message, ref, location=toctreenode)
-        raise
+        logger.warning(message, ref, location=toctreenode, type='toc', subtype=subtype)
+        raise DocumentError(message % ref) from err
```

Comment:
Why this change? We should probably keep the KeyError because I have no idea if it is caught somewhere else differently

Reply:
Only in [environment/__init__.py#L455](/sphinx-doc/sphinx/blame/v7.2.5/sphinx/environment/__init__.py#L455) when there is an OSError. 

`title = ref` and `nodes.reference(..., *[nodes.Text(title)])` in following lines might be only KeyError source. I think this DocumentError could be outstanding for users to check original document status or filesystem status.

Reply:
For now, I would keep the exception type because they might catch it or assume its type. Changing the exception type could lead to incompatible changes and this requires a major version.